KeeTrayTOTP.Tests/Base32Tests.cs

```diff
@@ -0,0 +1,109 @@
+﻿using FluentAssertions;
+using KeeTrayTOTP.Libraries;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace KeeTrayTOTP.Tests
+{
+    [TestClass]
+    public class Base32Tests
+    {
+        private const string InvalidSeed = "MZXW6!YTBOI======!";
+
+        [DataTestMethod]
+        [DynamicData(nameof(GetBase32Data), DynamicDataSourceType.Method)]
+        public void Encode_ShouldWorkOnRfc4648TestCases(Base32TestCase testCase)
+        {
+            var actual = Base32.Encode(Encoding.UTF8.GetBytes(testCase.Decoded));
+
+            actual.Should().Be(testCase.Encoded);
+        }
+
+        [DataTestMethod]
+        [DynamicData(nameof(GetBase32Data), DynamicDataSourceType.Method)]
+        public void Decode_ShouldWorkOnRfc4648TestCases(Base32TestCase testCase)
+        {
+            var actual = Base32.Decode(testCase.Encoded);
+
+            var actualString = Encoding.UTF8.GetString(actual);
+
+            actualString.Should().Be(testCase.Decoded);
+        }
+
+        [DataTestMethod]
+        [DynamicData(nameof(GetBase32Data), DynamicDataSourceType.Method)]
+        public void IsBase32Out_ShouldReturnTrueFor(Base32TestCase testCase)
+        {
+            var actual = testCase.Encoded.IsBase32(out string invalidChars);
+
+            actual.Should().BeTrue();
+            invalidChars.Should().BeNullOrEmpty();
+        }
+
+        [DataTestMethod]
+        [DynamicData(nameof(GetBase32Data), DynamicDataSourceType.Method)]
+        public void IsBase32_ShouldReturnTrueFor(Base32TestCase testCase)
+        {
+            var actual = testCase.Encoded.IsBase32();
+
+            actual.Should().BeTrue();
+        }
+
+        [TestMethod]
+        public void Decode_ShouldThrowWithInvalidSeed()
+        {
+            Action act = () => Base32.Decode(InvalidSeed);
+
+            act.Should().Throw<ArgumentException>().WithMessage("Base32 contains illegal characters.");
+        }
+
+        [TestMethod]
+        public void ExtIsBase32_ShouldReturnFalseWithInvalidChars()
+        {
+            var actual = InvalidSeed.IsBase32(out string invalidChars);
+
+            actual.Should().BeFalse();
+            invalidChars.Should().Be("!======!");
```

Comment:
And this should then only return ``!!`` as invalidChars, shouldn't it?

Reply:
This would depend on the outcome above. I think it could say just '!', or '!!', but in this example the = is not used as padding and is invalid as well...